linchpin/hooks/__init__.py

```diff
@@ -0,0 +1,200 @@
+        res_file = name+".output"
+        inv_file = "{0}/{1}".format(
+                   self.api.current_target_data["extra_vars"]["default_inventories_path"],
+                   inv_file
+                   )
+        res_file = "{0}/{1}".format(
+                   self.api.current_target_data["extra_vars"]["default_resources_path"],
+                   res_file
+                   )
+        self.api.current_target_data["extra_vars"]["inventory_file"] = inv_file
+        self.api.current_target_data["extra_vars"]["resource_file"] = res_file
+
+    def run_hooks(self, state, is_global=False):
+        
+        """
+        Get the cfgs object
+        :param section: section from ini-style config file
+        :param key: key to get from config file, within section
+        """
+
+        if not str(state) in self.api.ctx.cfgs["playbook_pre_states"]:
+            self.prepare_ctx_params()
+        self.api.ctx.log("State change triggered in linchpin API")
+        self.api.ctx.log("Observed State in LinchpinHooks :: "+str(state))
+        hooks_data = self.api.current_target_data.get("hooks", None)
+        if hooks_data == None:
+            self.api.ctx.log("No hooks found for current target")
+            return
+        # fetches all the state_data , ie., all the action blocks inside
+        # state of the target
+        state_data = hooks_data.get(str(state), None)
+
+        # Print out error message if the hooks are not found
+        if state_data == None:
+            self.api.ctx.log(str(state)+" State hook not found in PinFile")
+            return
+
+        # current target data extravars are fetched
+        target_data = self.api.current_target_data.get("extra_vars", None)
+        self.run_actions(state_data, target_data)
+
+    def run_actions(self, action_blocks, target_data, is_global=False):
+
+        """
+        Runs actions inside each action block of each target
+        params:
+        action_blocks: list of action_blocks each block constitues
+                       to a type of hook
+        target_data: data specific to target , which can be dict of topology , layout, outputs , inventory.
+        is_global: scope of the hook.
+        example: action_block: 
+        - name: do_something
+          type: shell
+          actions:
+            - echo " this is post up operation Hello hai how r u ?"
+        """
+
+        if is_global:
+            raise NotImplementedError("Run Hooks is not implemented \
+                                       for global scoped hooks")
+        else:
+            # a_b -> abbr for action_block
+            for a_b in action_blocks:
+                action_type = a_b["type"]
+                ctx = a_b['context'] if a_b.has_key('context') else True
+                if not a_b.has_key("path"):
+                    # if the path is not defined it defaults to
+                    # workspace/hooks/typeofhook/name
+                    a_b["path"] = "{0}/hooks/{1}/{2}/".format(
```

Comment:
hooks is hardcoded here. Yes, hooks will likely not change, but let's not use hardcoded values. Make it a configuration variable.